x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -554,10 +517,24 @@ private static void analyzeAndMaybeRetry(
         logicalPlanListener.onResponse(plan);
     }
 
-    static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchFields) {
+    private static void resolveFieldNames(LogicalPlan parsed, EnrichResolution enrichResolution, ActionListener<ListenerResult> l) {
+        try {
+            // we need the match_fields names from enrich policies and THEN, with an updated list of fields, we call field_caps API
+            var enrichMatchFields = enrichResolution.resolvedEnrichPolicies()
+                .stream()
+                .map(ResolvedEnrichPolicy::matchField)
+                .collect(Collectors.toSet());
+            // get the field names from the parsed plan combined with the ENRICH match fields from the ENRICH policy
+            l.onResponse(fieldNames(parsed, enrichMatchFields, new ListenerResult(enrichResolution)));
+        } catch (Exception ex) {
+            l.onFailure(ex);
+        }
+    }
```

Comment:
This code was there before, in the join.indices.size==0 case, and was specifically wrapping the listener.onResponse call. All other places where we call listener.onResponse we have a try/catch, so I think we should keep it. The one time we did see something being thrown and caught here was when one PR added behaviour to PlanNode.output() that relied on fields being resolved, but fieldNames() actually calls down to this, and fails if the fields are not yet resolved. This was fixed a couple of weeks back, but I can imagine future scenarios where this could happen.